java/src/test/java/ai/onnxruntime/OnnxTensorTest.java

```diff
@@ -88,6 +88,91 @@ public void testScalarCreation() throws OrtException {
     }
   }
 
+  @Test
+  public void testBufferCreation() throws OrtException {
+    OrtEnvironment env = OrtEnvironment.getEnvironment();
+
+    // Test creating a value from an array
+    // Arrays result in tensors allocated by ORT, so they do not have a backing java.nio.Buffer
+    float[] arrValues = new float[] {0, 1, 2, 3, 4};
+    try (OnnxTensor t = OnnxTensor.createTensor(env, arrValues)) {
+      // array creation isn't backed by buffers
+      Assertions.assertFalse(t.ownsBuffer());
+      Assertions.assertFalse(t.getBufferRef().isPresent());
+      FloatBuffer buf = t.getFloatBuffer();
+      float[] output = new float[arrValues.length];
+      buf.get(output);
+      Assertions.assertArrayEquals(arrValues, output);
+
+      // Can't modify the tensor through this buffer.
+      buf.put(0, 25);
+      Assertions.assertArrayEquals(arrValues, output);
+    }
+
+    // Test creating a value from a non-direct byte buffer
```

Comment:
I am not sure what is non-direct buffer. I thought if we utilize  Java nio buffers, we can pass those pointers to ORT directly. But here it says I needs to copy.

Reply:
A direct buffer is allocated in Java by calling `ByteBuffer.allocateDirect` and is a pointer to off-heap memory wrapped in a Java object. In JNI the pointer can be extracted and fed into ORT. A non-direct buffer is allocated with `ByteBuffer.allocate` or `FloatBuffer.allocate` etc, and is backed by a on heap Java array of the appropriate type. Unfortunately this isn't expressed in the type system in a visible way so I can't make it clearer to users. The `OnnxTensor.createTensor` methods document this - https://github.com/microsoft/onnxruntime/blob/main/java/src/main/java/ai/onnxruntime/OnnxTensor.java#L395.

Reply:
So there is not such a thing as managed buffer pinning.

Reply:
One other thing, once you've created this Tensor from a byte buffer of any kind on the Java side, it is then backed by a direct byte buffer that users can manipulate and overwrite the values in as they see fit, it's only the initial creation that triggers a copy from a non-direct byte buffer. Once I've moved the array code over to creating a buffer and copying into that internally then it will also expose a direct byte buffer which can be manipulated to update it with new values.